File: services/28-java-grpc/src/main/java/ca/yarova/grpc/security/JwtUtil.java
package ca.yarova.grpc.security;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;

import javax.crypto.SecretKey;
import java.nio.charset.StandardCharsets;
import java.util.Date;
import java.util.UUID;

public class JwtUtil {

    private static final long EXPIRY_MS = 8L * 60 * 60 * 1000; // 8 hours

    private final SecretKey signingKey;

    public JwtUtil(String secret) {
        // Pad secret to at least 256 bits for HS256
        String padded = secret.length() < 32
                ? secret + "0".repeat(32 - secret.length())
                : secret;
        this.signingKey = Keys.hmacShaKeyFor(padded.getBytes(StandardCharsets.UTF_8));
    }

    public static JwtUtil fromEnv() {
        String secret = System.getenv("JWT_SECRET");
        if (secret == null || secret.isBlank()) {
            throw new IllegalStateException("JWT_SECRET environment variable is not set");
        }
        return new JwtUtil(secret);
    }

    public String generateToken(UUID userId, String email, String name) {
        Date now = new Date();
        Date expiry = new Date(now.getTime() + EXPIRY_MS);

        return Jwts.builder()
                .subject(userId.toString())
                .claim("email", email)
                .claim("name", name)
                .issuedAt(now)
                .expiration(expiry)
                .signWith(signingKey)
                .compact();
    }

    public Claims validateToken(String token) {
        return Jwts.parser()
                .verifyWith(signingKey)
                .build()
                .parseSignedClaims(token)
                .getPayload();
    }
}

File: services/28-kotlin-grpc/src/main/kotlin/ca/yarova/grpc/security/JwtUtil.kt
package ca.yarova.grpc.security

import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import java.nio.charset.StandardCharsets
import java.util.Date
import java.util.UUID

// JwtUtil — issues and validates JWT tokens for gRPC service auth.
//
// JWT secret: loaded from JWT_SECRET env var at construction.
// Expiry: 8 hours from issuance.
// Algorithm: HS256.

object JwtUtil {

    private const val EXPIRY_MS: Long = 8L * 60 * 60 * 1000 // 8 hours

    private val signingKey by lazy {
        val secret = System.getenv("JWT_SECRET")
            ?: throw IllegalStateException("JWT_SECRET environment variable is not set")
        val padded = if (secret.length < 32) secret.padEnd(32, '0') else secret
        Keys.hmacShaKeyFor(padded.toByteArray(StandardCharsets.UTF_8))
    }

    fun generateToken(userId: UUID, email: String, name: String): String {
        val now = Date()
        val expiry = Date(now.time + EXPIRY_MS)
        return Jwts.builder()
            .subject(userId.toString())
            .claim("email", email)
            .claim("name", name)
            .issuedAt(now)
            .expiration(expiry)
            .signWith(signingKey)
            .compact()
    }

    fun validateToken(token: String): Claims =
        Jwts.parser()
            .verifyWith(signingKey)
            .build()
            .parseSignedClaims(token)
            .payload
}
